### ISSUE TITLE
Reason and Impact of `Flow:FTL Re-executed forward pass failed` Error

### ISSUE THREAD

**Issue by y0ke** [ripple/rippled#2929]
I'm running multiple mainnet rippled in different locations.

All of them using rippled v1.2.2 and configs are
```
[node_size]
tiny

[node_db]
type=RocksDB
path=/var/lib/rippled/db/rocksdb
open_files=2000
filter_bits=12
cache_mb=256
file_size_mb=8
file_size_mult=2
online_delete=2000
advisory_delete=0

[database_path]
/var/lib/rippled/db
[debug_logfile]
/var/log/rippled/debug.log

[ips]
r.ripple.com 51235
```

All of them produced same fatal error `FTL Re-executed forward pass failed` at almost same time yesterday.(2019-May-07 15:16, 2019-May-07 15:20:55 and 2019-May-07 17:22 UTC )

It's been around 8 months since I've started running these node though, This is the first time I see this error. 

They are syncing fully without restarting and working as expected as far as I checked though I'm wondering exact reason why.

**Recorded Log sample (timestamp in UTC)**
```
cat debug.log | grep "FTL Re-executed forward pass"
2019-May-07 15:16:43.361342780 Flow:FTL Re-executed forward pass failed
2019-May-07 15:16:44.377049096 Flow:FTL Re-executed forward pass failed
2019-May-07 15:16:44.379657323 Flow:FTL Re-executed forward pass failed
2019-May-07 15:16:47.263107591 Flow:FTL Re-executed forward pass failed
2019-May-07 15:16:47.275215738 Flow:FTL Re-executed forward pass failed
2019-May-07 15:16:47.286666596 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:55.886536124 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:55.895162735 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:55.901387236 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:55.910942523 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:55.924334340 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:55.939768297 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:55.957397216 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:55.968119170 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:55.985485061 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:55.997884984 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:56.015215580 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:56.024735059 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:56.027959856 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:56.036839859 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:56.042680548 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:56.045008881 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:56.073727672 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.894666248 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.897469483 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.900029643 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.901470766 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.902944345 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.904483390 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.905939928 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.908542406 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.911121477 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.913769537 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.916281673 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.917695664 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.919157402 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.920549330 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.921983868 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.924457374 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.927113047 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.928617005 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.930067769 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.932674060 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.934069106 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.936651826 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.938052830 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.939489906 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.940927336 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.942337773 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.943838127 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.945375780 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.946859514 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.948351747 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.949780318 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.952401728 Flow:FTL Re-executed forward pass failed
2019-May-07 15:20:58.953841994 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:01.769431703 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:01.794877503 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:01.813680302 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:01.821707729 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:01.827141772 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:01.853349894 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:01.871523775 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:01.886497520 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:01.893405412 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:01.914148635 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:01.930538735 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:01.935406628 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:01.940452674 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:01.960765341 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:01.977526353 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:01.992012235 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:01.997090209 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:02.002023616 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:02.006921578 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:02.028075314 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:02.032799303 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:02.048842417 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:02.053939680 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:02.059018014 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:02.064018397 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:02.069188938 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:02.074239895 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:02.094531621 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:02.099083587 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:02.111846521 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:02.115460169 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:02.126403029 Flow:FTL Re-executed forward pass failed
2019-May-07 15:21:02.129738553 Flow:FTL Re-executed forward pass failed
2019-May-07 17:22:33.075856974 Flow:FTL Re-executed forward pass failed
2019-May-07 17:22:35.474775659 Flow:FTL Re-executed forward pass failed
2019-May-07 17:22:35.530788219 Flow:FTL Re-executed forward pass failed
2019-May-07 17:22:35.581993473 Flow:FTL Re-executed forward pass failed
```

**Comment by y0ke** [comment]
I'm considering the root cause is invalid ledgers published by validators but I'm not sure

**Comment by seelabs** [comment]
Thank you for your report. We detected the same log message yesterday. The cause of the message was a path in the payment engine was going to deliver a zero amount, when the engine expected a non-zero amount. This was caused by an exceptionally small amount of value being passed through the path (on the order of 10^-80).

As a result of this, the payment failed and a message was logged. We are working on a fix for this to be included in future versions of rippled. In the meantime, it is safe to ignore this message.
